### PR TITLE
Add Bounce example to Haptic Sampler

### PR DESCRIPTION
**Pull request by jmylen**

### CODE REVIEW

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    val collisionData = updateCollisionData(transitionData)
+
+    // Execute a click vibration once the ball has been reset.
+    var hasVibratedForReset by remember { mutableStateOf(false) }
+    if (collisionData.collisionWithReset) {
+        if (!hasVibratedForReset) {
+            clickVibration(vibrator)
+            hasVibratedForReset = true
+        }
+    }
```
Comment (laissandrade) [inline Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt:110]: Can we reset hasVibratedForReset here in a else clause, like we do below for hasVibratedForBallContact, to keep this flag controlled by at a single place?

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+            Modifier
+                .fillMaxSize()
+                .noRippleClickable {
+                    if (transitionData.isAtStart) {
+                        ballState = BallState.End
+                    } else {
+                        // Reset the state, with a thud vibration to simulate bounce off the floor.
+                        ballState = BallState.Start
+                        bounceCount = 0
+                        hasVibratedForBallContact = false
```
Comment (laissandrade) [inline Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt:135]: Is this extra reset for hasVibratedForBallContact necessary, or will the else clause above already clear this value on reset?

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                        hasVibratedForReset = false
+                        thudVibration(vibrator)
+                    }
+                },
+        ) {
+
+            var instructionsText: String = "" // Don't display any instructions when bouncing.
+            if (transitionData.isAtStart) {
+                instructionsText = stringResource(R.string.bounce_tap_to_drop)
+            } else if (transitionData.isAtEnd || transitionData.isResetting) {
```
Comment (laissandrade) [inline Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt:145]: Do you mean !transitionData.isResetting here? Or do you want to show only after the end of the bounce animation?

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+            BallState.End -> BALL_END_POSITION
+        }
+    }
+    val isAtTargetState = transition.currentState === transition.targetState
+    return TransitionData(
+        ballOffsetY = ballOffsetY,
+        isAtStart = isAtTargetState && transition.currentState == BallState.Start,
+        isAtEnd = isAtTargetState && transition.currentState == BallState.End,
+        isBouncing = !isAtTargetState && transition.targetState == BallState.End,
+        isResetting = transition.currentState != BallState.Start && transition.targetState == BallState.Start,
```
Comment (laissandrade) [inline Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt:243]: why not !isAtTargetState && transition.targetState == BallState.Start, for consistency with isBouncing?

**Review by laissandrade — CHANGES_REQUESTED** [review]
Thanks!

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+import com.example.android.haptics.samples.ui.components.Screen
+import com.example.android.haptics.samples.ui.modifiers.noRippleClickable
+import com.example.android.haptics.samples.ui.theme.HapticSamplerTheme
+import kotlin.math.pow
+
+/**
+ * The two states of the animated ball. Start refers to the elevated state in which
+ * the ball drops from, and End once the ball is at rest on the "floor".
+ */
+private enum class BallState {
```
[the remark this answers is not in the document]
Comment (jmylen) [reply Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt]: I liked your suggestion to rename the enum to BallPosition! Done.

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+@Composable
+private fun BounceExampleScreen(messageToUser: String) {
+    var ballState by remember { mutableStateOf(BallState.Start) }
+    var bounceCount by remember { mutableStateOf(0) }
+
+    val vibrator = LocalContext.current.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
+
+    var transitionData = updateTransitionData(
+        ballState = ballState
```
[the remark this answers is not in the document]
Comment (jmylen) [reply Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt]: Think I fixed what you mean.

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt (hunk 2 above)
Comment (jmylen) [reply Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt]: Good call.

File: Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt (hunk 1 above)
Comment (jmylen) [reply Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt:110]: Good call. Makes sense.

**Review by laissandrade — APPROVED** [review]
Thanks!

**Pull request merged**